source/common/http/http2/protocol_constraints.cc

```diff
@@ -0,0 +1,123 @@
+    : stats_(stats), max_outbound_frames_(http2_options.max_outbound_frames().value()),
+      frame_buffer_releasor_([this]() { releaseOutboundFrame(); }),
+      max_outbound_control_frames_(http2_options.max_outbound_control_frames().value()),
+      control_frame_buffer_releasor_([this]() { releaseOutboundControlFrame(); }),
+      max_consecutive_inbound_frames_with_empty_payload_(
+          http2_options.max_consecutive_inbound_frames_with_empty_payload().value()),
+      max_inbound_priority_frames_per_stream_(
+          http2_options.max_inbound_priority_frames_per_stream().value()),
+      max_inbound_window_update_frames_per_data_frame_sent_(
+          http2_options.max_inbound_window_update_frames_per_data_frame_sent().value()) {}
+
+ProtocolConstraints::ReleasorProc
+ProtocolConstraints::incrementOutboundFrameCount(bool is_outbound_flood_monitored_control_frame) {
+  ++outbound_frames_;
+  if (is_outbound_flood_monitored_control_frame) {
+    ++outbound_control_frames_;
+  }
+  status_.Update(checkOutboundQueueLimits());
+  return is_outbound_flood_monitored_control_frame ? control_frame_buffer_releasor_
+                                                   : frame_buffer_releasor_;
+}
+
+void ProtocolConstraints::releaseOutboundFrame() {
+  ASSERT(outbound_frames_ >= 1);
+  --outbound_frames_;
+}
+
+void ProtocolConstraints::releaseOutboundControlFrame() {
+  ASSERT(outbound_control_frames_ >= 1);
+  --outbound_control_frames_;
+  releaseOutboundFrame();
+}
+
+Status ProtocolConstraints::checkOutboundQueueLimits() {
+  // Stop checking for further violations after the first failure.
+  if (!status_.ok()) {
+    return status_;
+  }
+
+  if (outbound_frames_ > max_outbound_frames_) {
+    stats_.outbound_flood_.inc();
+    return bufferFloodError("Too many frames in the outbound queue.");
+  }
+  if (outbound_control_frames_ > max_outbound_control_frames_) {
+    stats_.outbound_control_flood_.inc();
+    return bufferFloodError("Too many control frames in the outbound queue.");
+  }
+  return okStatus();
+}
+
+Status ProtocolConstraints::trackInboundFrames(const nghttp2_frame_hd* hd,
+                                               uint32_t padding_length) {
+  switch (hd->type) {
+  case NGHTTP2_HEADERS:
+  case NGHTTP2_CONTINUATION:
+    // Track new streams.
+    if (hd->flags & NGHTTP2_FLAG_END_HEADERS) {
+      inbound_streams_++;
+    }
+    FALLTHRU;
+  case NGHTTP2_DATA:
+    // Track frames with an empty payload and no end stream flag.
+    if (hd->length - padding_length == 0 && !(hd->flags & NGHTTP2_FLAG_END_STREAM)) {
+      consecutive_inbound_frames_with_empty_payload_++;
+    } else {
+      consecutive_inbound_frames_with_empty_payload_ = 0;
+    }
+    break;
+  case NGHTTP2_PRIORITY:
+    inbound_priority_frames_++;
+    break;
+  case NGHTTP2_WINDOW_UPDATE:
+    inbound_window_update_frames_++;
+    break;
+  default:
+    break;
+  }
+
+  status_.Update(checkInboundFrameLimits());
```

Comment:
We seem to be using status_ for both the inbound and outbound directions.  Is that a problem?

Reply:
This is by design. The idea is that only the first violation is recorded and the connection is disconnected at that point. In this case unwinding a deep stack and possibly triggering other errors would not mask the first problem.